File: kt/Minsk/minsk/src/main/kotlin/dev/phytolizer/minsk/analysis/binding/BoundAssignmentExpression.kt
package dev.phytolizer.minsk.analysis.binding

import kotlin.reflect.KClass

internal class BoundAssignmentExpression(val name: String, val expression: BoundExpression) : BoundExpression() {
    override val kind: BoundNodeKind
        get() = BoundNodeKind.AssignmentExpression
    override val type: KClass<out Any>
        get() = expression.type
}

File: nim/src/minsk/codeAnalysis/binding/expressions/boundAssignmentExpression.nim
import minsk/minskObject

import minsk/codeAnalysis/binding/[
  boundExpression,
  boundNodeKind,
]

type
  BoundAssignmentExpression* = ref object of BoundExpression
    name*: string
    expression*: BoundExpression

proc newBoundAssignmentExpression*(
  name: string,
  expression: BoundExpression
): BoundAssignmentExpression =
  new(result)
  result.name = name
  result.expression = expression

method kind*(self: BoundAssignmentExpression): BoundNodeKind =
  BoundNodeKind.AssignmentExpression

method ty*(self: BoundAssignmentExpression): MinskObjectKind =
  self.expression.ty
